File: Week_03/G20190343020319/LeetCode_33_0317.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        if(nums.size() == 0) {
            return -1;
        }
        return search( nums, target, 0, nums.size() - 1);
    }

     int search(vector<int>& nums, int target, int begin, int end) {
        if(nums[begin] == target) {
            return begin;
        } 

        if(end <= begin) {
            return -1;
        } 
        if(nums[(begin + end)/2] == target) {
            return (begin + end)/2;
        } 
        if(nums[begin] > target) {
            if(nums[(begin + end)/2] > target) {
                if(nums[(begin + end)/2] > nums[begin]) {
                    return search( nums, target, (begin + end) / 2 + 1, end);
                } else {
                    return search( nums, target, begin +1 , (begin + end) / 2 -1);
                }
            } else {
                return search( nums, target, (begin + end) / 2 + 1, end);
            }
        } else {
            if(nums[(begin + end)/2] > nums[begin]) {
                if(nums[(begin + end)/2] > target) {
                    return search( nums, target, begin +1 , (begin + end) / 2 -1);
                } else {
                    return search( nums, target, (begin + end) / 2 + 1, end);
                    
                }    
                
            } else {
                if(nums[(begin + end)/2] > target) {
                    return search( nums, target, (begin + end) / 2 + 1, end);
                } else {
                    return search( nums, target, begin +1 , (begin + end) / 2 -1);
                    
                }    
            }                    
        }
        return -1;
    }
};

File: Week_03/G20190343020317/LeetCode_33_0317.java
class Solution {
    public int search(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int num;
            if ((nums[mid] < nums[0]) == (target < nums[0])) {
                num = nums[mid];
            } else {
                num = target < nums[0] ? Integer.MIN_VALUE : Integer.MAX_VALUE;
            }
            if (num < target){
                lo = mid + 1;
            }
            else if (num > target){
                hi = mid - 1;
            }
            else
                return mid;
        }
        return -1;
    }
}